pkg/controllers/vdb/replicatedupgrade_reconciler.go

```diff
@@ -929,3 +968,176 @@ func (r *ReplicatedUpgradeReconciler) getSandboxPodFacts(ctx context.Context, do
+	r.Manager.traceActorReconcile(actor)
+	return actor.Reconcile(ctx, &ctrl.Request{})
+}
+
+// removeReplicaGroupA will remove the old main cluster
+func (r *ReplicatedUpgradeReconciler) removeReplicaGroupA(ctx context.Context) (ctrl.Result, error) {
+	// if the sandbox is still there or replica group A has removed, we skip removing the old main cluster
+	if r.VDB.GetSandboxStatus(r.sandboxName) != nil || !r.isGroupASubclusterInStatus() {
+		return ctrl.Result{}, nil
+	}
+
+	actor := MakeDBRemoveSubclusterReconciler(r.VRec, r.Log, r.VDB, r.PFacts[vapi.MainCluster].PRunner,
+		r.PFacts[vapi.MainCluster], r.Dispatcher)
+	r.Manager.traceActorReconcile(actor)
+	return actor.Reconcile(ctx, &ctrl.Request{})
+}
+
+// uninstallReplicaGroupANodes will drive uninstall logic for any nodes in replica group A.
+func (r *ReplicatedUpgradeReconciler) uninstallReplicaGroupANodes(ctx context.Context) (ctrl.Result, error) {
+	// if the sandbox is still there or replica group A has removed, we skip removing the old main cluster
+	if r.VDB.GetSandboxStatus(r.sandboxName) != nil || !r.isGroupASubclusterInStatus() {
+		return ctrl.Result{}, nil
+	}
+
+	actor := MakeUninstallReconciler(r.VRec, r.Log, r.VDB,
+		r.PFacts[vapi.MainCluster].PRunner, r.PFacts[vapi.MainCluster])
+	r.Manager.traceActorReconcile(actor)
+	return actor.Reconcile(ctx, &ctrl.Request{})
+}
+
+// deleteReplicaGroupASts will delete the statefulSet of replicate group A.
+func (r *ReplicatedUpgradeReconciler) deleteReplicaGroupASts(ctx context.Context) (ctrl.Result, error) {
+	// if the sandbox is still there or replica group A has removed, we skip removing the old main cluster
+	if r.VDB.GetSandboxStatus(r.sandboxName) != nil || !r.isGroupASubclusterInStatus() {
+		return ctrl.Result{}, nil
+	}
+
+	actor := MakeObjReconciler(r.VRec, r.Log, r.VDB, r.PFacts[vapi.MainCluster], ObjReconcileModeAll)
+	r.Manager.traceActorReconcile(actor)
+	return actor.Reconcile(ctx, &ctrl.Request{})
+}
+
+// renameReplicaGroupBFromVdb will rename the subclusters in promoted-sandbox/new-main-cluster to
+// match the ones in original main cluster
+func (r *ReplicatedUpgradeReconciler) renameReplicaGroupBFromVdb(ctx context.Context) (ctrl.Result, error) {
+	// if replica group A still exists or subclusters in replica group B have been renamed, we skip
+	// renaming the subclusters in replica group B
+	if r.countSubclustersForReplicaGroup(vmeta.ReplicaGroupAValue) > 0 ||
+		r.areGroupBSubclustersRenamed() {
+		return ctrl.Result{}, nil
+	}
+
+	initiator, found := r.PFacts[vapi.MainCluster].FindFirstPrimaryUpPodIP()
+	if !found {
+		r.Log.Info("Requeue because there are no primary UP nodes in main cluster to execute rename-subcluster operation")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	scNames := r.getSubclustersForReplicaGroup(vmeta.ReplicaGroupBValue)
+	for _, scName := range scNames {
+		newScName, found := r.scNameMap[scName]
+		// ignore the subcluster that has been renamed
+		if !found {
```

Comment:
Does not found mean it has already been renamed?

Reply:
Yes, if the subcluster in group B cannot be found in scNameMap, it means that subcluster is renamed or removed. However, we won't remove subclusters in group B in replicated upgrade. So, we know the subcluster is renamed at this point.